rust_src/src/data.rs

```diff
@@ -693,4 +697,81 @@ pub fn set_default(symbol: LispSymbolRef, value: LispObject) -> LispObject {
     value
 }
 
+extern "C" fn harmonize_variable_watchers(alias: LispObject, base_variable: LispObject) {
+    if !base_variable.eq(alias)
+        && base_variable.eq(alias.as_symbol_or_error().get_indirect_variable().into())
+    {
+        alias
+            .as_symbol_or_error()
+            .set_trapped_write(base_variable.as_symbol_or_error().get_trapped_write());
+    }
+}
+
+/// Cause WATCH-FUNCTION to be called when SYMBOL is set.
+///
+/// It will be called with 4 arguments: (SYMBOL NEWVAL OPERATION WHERE).
+/// SYMBOL is the variable being changed.
+/// NEWVAL is the value it will be changed to.
+/// OPERATION is a symbol representing the kind of change, one of: `set',
+/// `let', `unlet', `makunbound', and `defvaralias'.
+/// WHERE is a buffer if the buffer-local value of the variable being
+/// changed, nil otherwise.
+///
+/// All writes to aliases of SYMBOL will call WATCH-FUNCTION too.
+#[lisp_fn]
+pub fn add_variable_watcher(symbol: LispSymbolRef, watch_function: LispObject) {
+    let symbol = symbol.get_indirect_variable();
+
+    symbol.set_trapped_write(symbol_trapped_write::SYMBOL_TRAPPED_WRITE);
+
+    map_obarray(
+        unsafe { globals.Vobarray },
+        harmonize_variable_watchers,
+        symbol.into(),
+    );
+
+    let watchers = unsafe { Fget(symbol.into(), Qwatchers) };
+    let mem = member(watch_function, watchers);
+
+    if mem.is_nil() {
+        put(symbol, Qwatchers, unsafe {
```

Comment:
I think you can use `LispObject::cons` here to avoid the `unsafe`.

Reply:
Is that something we should be doing everywhere?

Reply:
More Rust, Less C. Right?